parquet/src/main/java/org/apache/iceberg/parquet/vectorized/VectorizedParquetReader.java

```diff
@@ -0,0 +1,290 @@
+      }
+    }
+
+    ReadConf<T> copy() {
+      return new ReadConf<>(this);
+    }
+  }
+
+  private ReadConf conf = null;
+
+  private ReadConf init() {
+    if (conf == null) {
+      ReadConf readConf = new ReadConf(
+          input, options, expectedSchema, filter, batchReaderFunc, reuseContainers, caseSensitive, batchSize);
+      this.conf = readConf.copy();
+      return readConf;
+    }
+
+    return conf;
+  }
+
+  @Override
+  public Iterator iterator() {
+    FileIterator iter = new FileIterator(init());
+    addCloseable(iter);
+    return iter;
+  }
+
+  private static class FileIterator<T> implements Iterator<T>, Closeable {
+    private final ParquetFileReader reader;
+    private final boolean[] shouldSkip;
+    private final VectorizedReader<T> model;
+    private final long totalValues;
+    private final int batchSize;
+
+    private int nextRowGroup = 0;
+    private long nextRowGroupStart = 0;
+    private long valuesRead = 0;
+    private T last = null;
+
+    FileIterator(ReadConf conf) {
+      this.reader = conf.reader();
+      this.shouldSkip = conf.shouldSkip();
+      this.model = conf.model();
+      this.totalValues = conf.totalValues();
+      this.model.reuseContainers(conf.reuseContainers());
+      this.batchSize = conf.batchSize();
+    }
+
+    @Override
+    public boolean hasNext() {
+      return valuesRead < totalValues;
+    }
+
+    @Override
+    public T next() {
+      if (!hasNext()) {
+        throw new NoSuchElementException();
+      }
+      if (valuesRead >= nextRowGroupStart) {
+        advance();
+      }
+      this.last = model.read();
+      valuesRead += Math.min(nextRowGroupStart - valuesRead, batchSize);
+      return last;
+    }
+
+    private void advance() {
+      while (shouldSkip[nextRowGroup]) {
+        nextRowGroup += 1;
+        reader.skipNextRowGroup();
+      }
+
+      PageReadStore pages;
+      DictionaryPageReadStore dictionaryPageReadStore;
+      try {
+        dictionaryPageReadStore = reader.getNextDictionaryReader();
+        pages = reader.readNextRowGroup();
+      } catch (IOException e) {
+        throw new RuntimeIOException(e);
+      }
+
+      nextRowGroupStart += pages.getRowCount();
+      nextRowGroup += 1;
+      model.setRowGroupInfo(
+          pages,
+          dictionaryPageReadStore,
+          dictionaryPageReadStore == null ? null : buildColumnDictEncodedMap(reader.getRowGroups()));
+    }
+
+    /**
+     * Retuns a map of {@link ColumnPath} -> whether all the pages in the row group for this column are dictionary
+     * encoded
+     */
+    private static Map<ColumnPath, Boolean> buildColumnDictEncodedMap(List<BlockMetaData> blockMetaData) {
+      Map<ColumnPath, Boolean> map = new HashMap<>();
+      for (BlockMetaData b : blockMetaData) {
+        for (ColumnChunkMetaData c : b.getColumns()) {
+          map.put(c.getPath(), !ParquetUtil.hasNonDictionaryPages(c));
+        }
+      }
+      return map;
```

Comment:
The effect of this code is to use the result of `hasNonDictionaryPages` for just the last block because the value is replaced each time. I think this should AND the results together so that if any row group has non-dictionary pages, all batches are eagerly decoded.

Reply:
Fixed